docs: how to write a changeset (#4654)
biomejs/biome: CONTRIBUTING.md

### Changelog

This repository uses [knope](https://knope.tech/) to automate the releases of Biome's binaries, the Rust crates, the JavaScript libraries and the creation of the `CHANGELOG.md` for each library/crate.

If the PR you're about to open is a bugfix/feature visible to Biome users, you are encouraged to provide a **changeset** . To *create* a changeset, use the following command (*don't create it manually*):

```shell
just new-changeset
```
The command will present a prompt where you need to choose the libraries involved by the PR, the type of change (`major`, `minor` or `patch`) for each library, and a description of the change. The description will be used as name of the file.

The command will create the changeset(s) in the `.changeset` folder. You're free to open the file, and add more information in it.

#### Choose the correct packages

> In the vast majority of cases, you want to choose the `cli` library, which represent the main binary (AKA the npm package `@biomejs/biome`). If your PR also changes any *public* crate that is published, you'll have to select it too.

Reviewer: It sounds a bit counterintuitive to call `cli` a library. Is that just Knope terminology here? Otherwise I’d suggest “component”.

Author: This name is arbitrary, but it will be used also in the git tags as prefix, and that's currently the nomenclature we use.

Changing it would mean to update `setup-biome` and other editor extensions to use a new nomenclature, which relies on those prefixes.

I opted for a name less meaningful, but less breaking changes to the bigger ecosystem

Reviewer: Maybe we’re talking about something else? I didn’t mean it’s weird that `cli` is a component, just that the word “library” doesn’t really apply to it.

Author: Knope calls them "package", but in our case that `cli` is a black sheep because it's linked to multiple packages.

I don't mind using the word component

> In the vast majority of cases, you want to choose the `cli` library, which represent the main binary (AKA the npm package `@biomejs/biome`). If your PR also changes any *public* crate that is published, you'll have to select it too.

Reviewer 2: ```suggestion
In the vast majority of cases, you want to choose the `cli` library, which represents the main binary (AKA the npm package `@biomejs/biome`). If your PR also changes any *public* crate that is published, you'll have to select it too.
```

> In the vast majority of cases, you want to choose the `cli` library, which represent the main binary (AKA the npm package `@biomejs/biome`). If your PR also changes any *public* crate that is published, you'll have to select it too.

Author: That's covered in the example below, or were you thinking of something else?

> In the vast majority of cases, you want to choose the `cli` library, which represent the main binary (AKA the npm package `@biomejs/biome`). If your PR also changes any *public* crate that is published, you'll have to select it too.

Reviewer 3: Oh I see what you mean. You're talking about if a change touches a crate API and a user-facing thing like a lint rule?

Exactly :)

For example, if your PR fixes a *JavaScript* lint rule and changes a public of the crate `biome_js_syntax`, you'll select:
- `cli`
- `biome_js_analyze`
- `biome_js_syntax`

#### Choose the correct type of change

We are very strict about `major` changes in the `cli` package. To better understand type of your change *for this package*, please refer to our [versioning page](https://biomejs.dev/internals/versioning/). Generally:
- `patch`: any sort of change that fixes a bug.
- `minor`: new features available to the users.
- `major`: a change that breaks a user API.

> We are very liberal for the `biome_` crates, so don't be afraid to break the user face APIs, as long as the change is properly described.

Reviewer: ```suggestion
We are very liberal for the `biome_` crates, so don't be afraid to break the developer-facing APIs, as long as the change is properly described.
```

#### Writing a changeset

`knope` allows to generate two types of changesets, and you can use the one that you see fit:
1. A changeset with a single header (`#`). When there's a single header, `knope` will transform it into a single bullet point inside the `CHANGELOG.md`.

> 2. A changeset wit ha single header (`#`) and a paragraph beneath. `knope` will transform it into specialised header inside the `CHANGELOG.md`.

Reviewer: ```suggestion
2. A changeset with a single header (`#`) and a paragraph beneath. `knope` will transform it into a specialised header inside the `CHANGELOG.md`.
```

For example, you have a PR that ships a bug fix, and a new feature, and you want to document both of them. The bugfix doesn't require a long description, while the feature might need some explanation. So create two changesets, that might look like this:

> This the changeset for the bugfix:

Reviewer: ```suggestion
This is the changeset for the bugfix:
```

```markdown 
---
biome_js_syntax: patch
---

# Fix [#000](https://path/to/000), where the parameter `foo` wasn't read by the function `apply_foo()` 
```

> This is the changeset for the feature, where show the feature usage by having an example of configuration, with the shell result:

Reviewer: ```suggestion
This is the changeset for the feature, where you show the feature usage by having an example of configuration, with the shell result:
```

``````markdown
---
cli: minor
---

# Add the X feature

This feature does wonders, and you can use it with this configuration:

```json
{
    "configuration": "here"
}
```

And this happens: 

```shell

```
``````

These two changesets will be compiled by `knope`, and the final result in the `CHANGELOG.md` will be as follows:

``````markdown
### Patches

- Fix [#000](https://path/to/000), where the parameter `foo` wasn't read by the function `apply_foo()`

### Features

#### Add the X feature

This feature does wonders, and you can use it with this configuration:

```json
{
    "configuration": "here"
}
```

And this happens:

```shell

```
``````

Regarding the description of the changeset, try to follow the these guidelines:

- Use the present tense, e.g. "Add new feature", "Fix edge case".
- If you fix a bug, please add the link to the issue, e.g. "Fix edge case [#4444]()".
- Whenever applicable, add a code block to show your new changes. For example, for a new
  rule you might want to show an invalid case, for the formatter you might want to show
  how the new formatting changes, and so on.

> If in doubt, take a look to existing or past changesets.

Reviewer: ```suggestion
If in doubt, take a look at existing or past changesets.
```
